pkg/segment/writer/segmetarw.go

```diff
@@ -480,12 +492,107 @@ func BackFillPQSSegmetaEntry(segkey string, newpqid string) {
 	if sfmData.AllPQIDs == nil {
 		sfmData.AllPQIDs = make(map[string]bool)
 	}
-	sfmData.AllPQIDs[newpqid] = true
 
-	// TODO this should be through a channel so that we only have one writer
+	utils.MergeMapsRetainingFirst(sfmData.AllPQIDs, pqidMap)
+
 	writeSfm(segkey, sfmData)
 }
 
+func BackFillPQSSegmetaEntry(segkey string, newpqid string) {
+	BulkBackFillPQSSegmetaEntries(segkey, map[string]bool{newpqid: true})
+}
+
+// AddToBackFillAndEmptyPQSChan adds a new pqid to the channel
+// if emptyPqs is true, then it will also add the EmptyResults for this pqid
+func AddToBackFillAndEmptyPQSChan(segkey string, newpqid string, emptyPqs bool) {
+	pqsChan <- PQSChanMeta{pqid: newpqid, segKey: segkey, emptyPqs: emptyPqs}
+}
+
+func listenBackFillAndEmptyPQSRequests() {
+	// Listen on the channel, every 10 seconds or if the size of the channel is 100,
+	// it would get all the data in the channel and then do the process of Backfilling PQMR files.
+	// This is to avoid multiple writes to the same file
+
+	ticker := time.NewTicker(10 * time.Second) // every 10 seconds
+	defer ticker.Stop()
+
+	buffer := make([]PQSChanMeta, 0, 100)
+	var processing int32
+
+	isProcessing := func() bool {
+		return atomic.LoadInt32(&processing) == 1
+	}
+
+	callProcessBackFillPQSRequests := func() {
+		atomic.StoreInt32(&processing, 1)
+		bufferCopy := make([]PQSChanMeta, len(buffer))
+		copy(bufferCopy, buffer)
+		buffer = buffer[:0]
+		go func(pqsRequests []PQSChanMeta) {
+			processBackFillAndEmptyPQSRequests(pqsRequests)
+			atomic.StoreInt32(&processing, 0)
+		}(bufferCopy)
+	}
+
+	for {
+		select {
+		case pqsChanMeta := <-pqsChan:
+			buffer = append(buffer, pqsChanMeta)
+			if len(buffer) >= 100 && !isProcessing() {
+				callProcessBackFillPQSRequests()
```

Comment:
see comment above, we should accumulate upto x things and then call the process func in sync, and then return back to the select loop.

Reply:
nit: 
1)we are calling a one-liner function `callProcessBackFillPQSRequests` and resetting the buf len. We can just call the `processBackFillAndEmptyPQSRequests` directly

2) also do we need a append call here ? we could use a counter and pass the counter to the processBackfill func, so that it knows what is the length. But if you like append, you can keep that

Reply:
1. Yes, we could just call there instead of calling `callProcessBackFillPQSRequests`. I have made this function because, I was first trying to synchronize to keep the async of receiving data from the channel and made it a non-blocking one. So there was similar code in both the cases and it made sense for me to just make this `callProcessBackFillPQSRequests` internal function which both the cases will call.
   - But when I changed it to blocking one, I still kept this because, we will still be rewriting the reset of buffer line in both cases and may be if we add some other functionality then it might be better to have this single function that handles the process. 
   - But I will change it to how you have suggested for better readability.

2. Yes, that makes sense. I used append, and initially created the slice with the capacity of `100` so that the buffer could also handle more than `100` items during the non-blocking version. But since, it is a blocking one, I could change to the way you have suggested.